Source/JavaScriptCore/yarr/YarrInterpreter.cpp

```diff
@@ -408,47 +490,76 @@ class Interpreter {
+            return false;
+
+        return term.atom.patternCharacter == input.tryReadBackward(negativeInputOffset);
     }
 
-    bool checkSurrogatePair(int testUnicodeChar, unsigned negativeInputOffset)
+    bool checkSurrogatePair(ByteTerm& term, unsigned negativeInputOffset)
     {
-        return testUnicodeChar == input.readSurrogatePairChecked(negativeInputOffset);
+        ASSERT(term.isCharacterType());
+        if (term.matchDirection() == Forward)
+            return term.atom.patternCharacter == input.readSurrogatePairChecked(negativeInputOffset);
+
+        return term.atom.patternCharacter == input.readBackwardSurrogatePairChecked(negativeInputOffset);
+
     }
 
-    bool checkCasedCharacter(int loChar, int hiChar, unsigned negativeInputOffset)
+    bool checkCasedCharacter(ByteTerm& term, unsigned negativeInputOffset)
     {
-        int ch = input.readChecked(negativeInputOffset);
-        return (loChar == ch) || (hiChar == ch);
+        ASSERT(term.isCasedCharacterType());
+
+        int ch = term.matchDirection() == Forward ? input.readChecked(negativeInputOffset) : input.tryReadBackward(negativeInputOffset);
+        return (term.atom.casedCharacter.lo == ch) || (term.atom.casedCharacter.hi == ch);
     }
 
-    bool checkCharacterClass(CharacterClass* characterClass, bool invert, unsigned negativeInputOffset)
+    bool checkCharacterClass(ByteTerm& term, unsigned negativeInputOffset)
     {
-        int inputChar = input.readChecked(negativeInputOffset);
+        ASSERT(term.isCharacterClass());
+
+        int inputChar = term.matchDirection() == Forward ? input.readChecked(negativeInputOffset) : input.tryReadBackward(negativeInputOffset);
         if (inputChar < 0)
             return false;
 
-        bool match = testCharacterClass(characterClass, inputChar);
-        return invert ? !match : match;
+        bool match = testCharacterClass(term.atom.characterClass, inputChar);
+        return term.invert() ? !match : match;
     }
     
-    bool checkCharacterClassDontAdvanceInputForNonBMP(CharacterClass* characterClass, unsigned negativeInputOffset)
+    bool checkCharacterClassDontAdvanceInputForNonBMP(ByteTerm& term, unsigned negativeInputOffset)
     {
-        int readCharacter = characterClass->hasOnlyNonBMPCharacters() ? input.readSurrogatePairChecked(negativeInputOffset) :  input.readChecked(negativeInputOffset);
+        ASSERT(term.isCharacterClass());
+        CharacterClass* characterClass = term.atom.characterClass;
+
+        int readCharacter = -1;
+
+        if (term.matchDirection() == Forward)
+            readCharacter = characterClass->hasOnlyNonBMPCharacters() ? input.readSurrogatePairChecked(negativeInputOffset) :  input.readChecked(negativeInputOffset);
+        else
+            readCharacter = characterClass->hasOnlyNonBMPCharacters() ? input.readBackwardSurrogatePairChecked(negativeInputOffset) :  input.tryReadBackward(negativeInputOffset);
```

Comment:
Ditto. And if we need to additional check for backwards, then why is `readBackwardSurrogatePairChecked` paired with `readSurrogatePairChecked`? Shouldn't it be `tryReadBackwardSurrogatePair`?

Reply:
`readBackwardSurrogatePairChecked()` is no longer used.  I deleted it.

Reply:
OK, so I will see after the patch is updated.